packages/worker/src/api/controllers/global/self.ts

```diff
@@ -123,11 +123,12 @@ export async function updateSelf(
   ctx: UserCtx<UpdateSelfRequest, UpdateSelfResponse>
 ) {
   const body = ctx.request.body
-  const update: UpdateSelf = {
-    firstName: body.firstName,
-    lastName: body.lastName,
-    password: body.password,
-    forceResetPassword: body.forceResetPassword,
+
+  const update: UpdateSelf = {}
+  for (let [key, value] of Object.entries(body)) {
+    if (value) {
+      update[key as keyof UpdateSelf] = value
+    }
   }
```

Comment:
Hey Rory, thanks for taking a look! This was purely to avoid accidentally nullifying any of the permitted fields in the update. Instead of passing the entire user up with the request now, we're only passing the values we explicitly wish to change. If I update just the `onboardedAt` in the updateSelf request I would end up deleting any of the other 4 values from my user when submitted.
The for loop was just a way to ensure empty values were completely omitted from any update attempt.

Reply:
In that case given that what we are aiming at is preventing bad requests from being sent, could we use a more appropriate validator?
```ts
  schema = {
    password: Joi.string().optional(),
    forceResetPassword: Joi.boolean().optional(),
    firstName: Joi.string().allow("").optional(),
    lastName: Joi.string().allow("").optional(),
    onboardedAt: Joi.string().optional(),
  }
```
where each field is optional, and `null` is not allowed

Reply:
That's definitely a much nicer solution for the validator but valid requests would still cause issues without something like the for loop in place. If I updated only the `onboardedAt` value, the update object would end up like this even with the validator present:

```javascript
update = {
    firstName: undefined,
    lastName: undefined,
    password: undefined,
    forceResetPassword: undefined,
    onboardedAt: '2023-03-07T15:13:12.546Z',
}
```

When this is passed to the spread operation in the users sdk, the undefined values are stripped from the user entirely 
https://github.com/Budibase/budibase/blob/master/packages/worker/src/sdk/users/users.ts#L232

The above update would the clear the 4 other elements from the user entirely with the password only being restored inside the save operation.

Reply:
We can switch out the spread in that case for the lodash `merge` function in that case, as:
```ts
import * as _ from "lodash"

_.merge(user, data)
```

Reply:
with the spread and the validation I think it's works as intended? this would be without the need for any lodash merging 

![Screenshot 2023-03-07 at 16 02 45](https://user-images.githubusercontent.com/8755148/223478082-85c0a5f8-4b1e-416f-9970-c3fd448d5f54.png)

Reply:
I had a look at this - added a little backend-core utility to make this a ton easier:
```typescript
  const update: UpdateSelf = {
    ...apiUtils.cleanRequestBody<UpdateSelfRequest>(body),
  }
```

Essentially - it uses https://www.npmjs.com/package/ts-transformer-keys to get the keys of the `UpdateSelfRequest` and then makes sure that the body only contains those keys - gives the best of both worlds, allows dynamic spreading like this without the risk of carrying over unexpected properties.